File: packages/sdk/java/src/main/java/com/ihui/ai/sdk/module/KnowledgeApi.java
package com.ihui.ai.sdk.module;

import com.fasterxml.jackson.databind.JsonNode;
import com.ihui.ai.sdk.BaseClient;
import com.ihui.ai.sdk.model.KnowledgeSearchRequest;

import java.util.List;
import java.util.Map;

/**
 * 知识库 / RAG / 知识图谱模块。
 *
 * <p>端点(13 个):
 * <ul>
 *   <li>GET    /v1/knowledge/health</li>
 *   <li>GET    /v1/knowledge/documents</li>
 *   <li>POST   /v1/knowledge/documents</li>
 *   <li>GET    /v1/knowledge/documents/:id</li>
 *   <li>GET    /v1/knowledge/documents/:id/chunks</li>
 *   <li>DELETE /v1/knowledge/documents/:id</li>
 *   <li>POST   /v1/knowledge/search</li>
 *   <li>POST   /v1/knowledge/rag-context</li>
 *   <li>POST   /v1/knowledge-graph/extract</li>
 *   <li>POST   /v1/knowledge-graph/build</li>
 *   <li>GET    /v1/knowledge-graph/data</li>
 *   <li>DELETE /v1/knowledge-graph/data</li>
 *   <li>POST   /v1/knowledge/documents/batch-delete</li>
 * </ul>
 */
public final class KnowledgeApi {

    private final BaseClient client;

    /**
     * 构造 KnowledgeApi。
     *
     * @param client 底层 BaseClient
     */
    public KnowledgeApi(BaseClient client) {
        this.client = client;
    }

    /**
     * GET /v1/knowledge/health(健康检查)。
     *
     * @return 健康状态(JsonNode,含 status / documents / chunks)
     */
    public JsonNode health() {
        return client.request("GET", "/knowledge/health", null, JsonNode.class);
    }

    /**
     * GET /v1/knowledge/documents(文档列表)。
     *
     * @return 文档列表(JsonNode)
     */
    public JsonNode listDocuments() {
        return client.request("GET", "/knowledge/documents", null, JsonNode.class);
    }

    /**
     * POST /v1/knowledge/documents(文档入库)。
     *
     * @param req 请求体(title / content / source 等)
     * @return 入库结果(JsonNode)
     */
    public JsonNode ingestDocument(Map<String, Object> req) {
        return client.request("POST", "/knowledge/documents", req, JsonNode.class);
    }

    /**
     * GET /v1/knowledge/documents/:id(文档详情)。
     *
     * @param id 文档 ID
     * @return 文档详情(JsonNode)
     */
    public JsonNode getDocument(String id) {
        return client.request("GET", "/knowledge/documents/" + BaseClient.encode(id),
                null, JsonNode.class);
    }

    /**
     * GET /v1/knowledge/documents/:id/chunks(文档分块)。
     *
     * @param id 文档 ID
     * @return 分块列表(JsonNode)
     */
    public JsonNode getDocumentChunks(String id) {
        return client.request("GET",
                "/knowledge/documents/" + BaseClient.encode(id) + "/chunks",
                null, JsonNode.class);
    }

    /**
     * DELETE /v1/knowledge/documents/:id(删除文档)。
     *
     * @param id 文档 ID
     */
    public void deleteDocument(String id) {
        client.requestRaw("DELETE", "/knowledge/documents/" + BaseClient.encode(id), null);
    }

    /**
     * POST /v1/knowledge/search(语义搜索)。
     *
     * @param req 搜索请求
     * @return 搜索结果(JsonNode,含 data 数组)
     */
    public JsonNode search(KnowledgeSearchRequest req) {
        return client.request("POST", "/knowledge/search", req, JsonNode.class);
    }

    /**
     * POST /v1/knowledge/rag-context(RAG 上下文检索)。
     *
     * @param req 请求体(query / topK 等)
     * @return RAG 上下文(JsonNode)
     */
    public JsonNode ragContext(Map<String, Object> req) {
        return client.request("POST", "/knowledge/rag-context", req, JsonNode.class);
    }

    /**
     * POST /v1/knowledge-graph/extract(知识图谱抽取)。
     *
     * @param req 请求体(text / documentId 等)
     * @return 抽取结果(JsonNode)
     */
    public JsonNode extractGraph(Map<String, Object> req) {
        return client.request("POST", "/knowledge-graph/extract", req, JsonNode.class);
    }

    /**
     * POST /v1/knowledge-graph/build(知识图谱构建)。
     *
     * @param req 请求体(source / sourceType 等)
     * @return 构建结果(JsonNode)
     */
    public JsonNode buildGraph(Map<String, Object> req) {
        return client.request("POST", "/knowledge-graph/build", req, JsonNode.class);
    }

    /**
     * GET /v1/knowledge-graph/data(知识图谱数据)。
     *
     * @return 图谱数据(JsonNode,含 nodes / edges)
     */
    public JsonNode getGraphData() {
        return client.request("GET", "/knowledge-graph/data", null, JsonNode.class);
    }

    /**
     * DELETE /v1/knowledge-graph/data(清空知识图谱)。
     */
    public void clearGraph() {
        client.requestRaw("DELETE", "/knowledge-graph/data", null);
    }

    /**
     * POST /v1/knowledge/documents/batch-delete(批量删除)。
     *
     * @param documentIds 文档 ID 列表
     * @return 删除结果(JsonNode,含 deleted 计数)
     */
    public JsonNode batchDeleteDocuments(List<String> documentIds) {
        return client.request("POST", "/knowledge/documents/batch-delete",
                java.util.Collections.singletonMap("documentIds", documentIds), JsonNode.class);
    }
}

File: packages/sdk/dotnet/Modules/KnowledgeApi.cs
using System.Text.Json;

namespace Ihui.AI;

/// <summary>
/// 知识库 / RAG / 知识图谱模块。
/// </summary>
/// <remarks>
/// 端点(13 个):
/// <list type="bullet">
///   <item>GET    /v1/knowledge/health</item>
///   <item>GET    /v1/knowledge/documents</item>
///   <item>POST   /v1/knowledge/documents</item>
///   <item>GET    /v1/knowledge/documents/:id</item>
///   <item>GET    /v1/knowledge/documents/:id/chunks</item>
///   <item>DELETE /v1/knowledge/documents/:id</item>
///   <item>POST   /v1/knowledge/search</item>
///   <item>POST   /v1/knowledge/rag-context</item>
///   <item>POST   /v1/knowledge/graph/extract</item>
///   <item>POST   /v1/knowledge/graph/build</item>
///   <item>GET    /v1/knowledge/graph/data</item>
///   <item>DELETE /v1/knowledge/graph/data</item>
///   <item>POST   /v1/knowledge/documents/batch-delete</item>
/// </list>
/// </remarks>
public sealed class KnowledgeApi
{
    private readonly BaseClient _client;

    internal KnowledgeApi(BaseClient client) => _client = client;

    /// <summary>GET /v1/knowledge/health(健康检查)。</summary>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>健康状态(JsonElement,含 status / documents / chunks)</returns>
    public Task<JsonElement?> HealthAsync(CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("GET", "/knowledge/health", null, cancellationToken);

    /// <summary>GET /v1/knowledge/documents(文档列表)。</summary>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>文档列表(JsonElement)</returns>
    public Task<JsonElement?> ListDocumentsAsync(CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("GET", "/knowledge/documents", null, cancellationToken);

    /// <summary>POST /v1/knowledge/documents(文档入库)。</summary>
    /// <param name="req">请求体(title / content / source 等)</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>入库结果(JsonElement)</returns>
    public Task<JsonElement?> IngestDocumentAsync(object req, CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("POST", "/knowledge/documents", req, cancellationToken);

    /// <summary>GET /v1/knowledge/documents/:id(文档详情)。</summary>
    /// <param name="id">文档 ID</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>文档详情(JsonElement)</returns>
    public Task<JsonElement?> GetDocumentAsync(string id, CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("GET", "/knowledge/documents/" + BaseClient.Encode(id), null, cancellationToken);

    /// <summary>GET /v1/knowledge/documents/:id/chunks(文档分块)。</summary>
    /// <param name="id">文档 ID</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>分块列表(JsonElement)</returns>
    public Task<JsonElement?> GetDocumentChunksAsync(string id, CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("GET", "/knowledge/documents/" + BaseClient.Encode(id) + "/chunks", null, cancellationToken);

    /// <summary>DELETE /v1/knowledge/documents/:id(删除文档)。</summary>
    /// <param name="id">文档 ID</param>
    /// <param name="cancellationToken">取消令牌</param>
    public Task DeleteDocumentAsync(string id, CancellationToken cancellationToken = default)
        => _client.RequestRawAsync("DELETE", "/knowledge/documents/" + BaseClient.Encode(id), null, cancellationToken)!;

    /// <summary>POST /v1/knowledge/search(语义搜索)。</summary>
    /// <param name="req">搜索请求</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>搜索结果(JsonElement,含 data 数组)</returns>
    public Task<JsonElement?> SearchAsync(KnowledgeSearchRequest req, CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("POST", "/knowledge/search", req, cancellationToken);

    /// <summary>POST /v1/knowledge/rag-context(RAG 上下文检索)。</summary>
    /// <param name="req">请求体(query / topK 等)</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>RAG 上下文(JsonElement)</returns>
    public Task<JsonElement?> RagContextAsync(object req, CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("POST", "/knowledge/rag-context", req, cancellationToken);

    /// <summary>POST /v1/knowledge/graph/extract(知识图谱抽取)。</summary>
    /// <param name="req">请求体(text / documentId 等)</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>抽取结果(JsonElement)</returns>
    public Task<JsonElement?> ExtractGraphAsync(object req, CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("POST", "/knowledge/graph/extract", req, cancellationToken);

    /// <summary>POST /v1/knowledge/graph/build(知识图谱构建)。</summary>
    /// <param name="req">请求体(source / sourceType 等)</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>构建结果(JsonElement)</returns>
    public Task<JsonElement?> BuildGraphAsync(object req, CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("POST", "/knowledge/graph/build", req, cancellationToken);

    /// <summary>GET /v1/knowledge/graph/data(知识图谱数据)。</summary>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>图谱数据(JsonElement,含 nodes / edges)</returns>
    public Task<JsonElement?> GetGraphDataAsync(CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("GET", "/knowledge/graph/data", null, cancellationToken);

    /// <summary>DELETE /v1/knowledge/graph/data(清空知识图谱)。</summary>
    /// <param name="cancellationToken">取消令牌</param>
    public Task ClearGraphAsync(CancellationToken cancellationToken = default)
        => _client.RequestRawAsync("DELETE", "/knowledge/graph/data", null, cancellationToken)!;

    /// <summary>POST /v1/knowledge/documents/batch-delete(批量删除)。</summary>
    /// <param name="documentIds">文档 ID 列表</param>
    /// <param name="cancellationToken">取消令牌</param>
    /// <returns>删除结果(JsonElement,含 deleted 计数)</returns>
    public Task<JsonElement?> BatchDeleteDocumentsAsync(List<string> documentIds, CancellationToken cancellationToken = default)
        => _client.RequestAsync<JsonElement>("POST", "/knowledge/documents/batch-delete",
            new { documentIds }, cancellationToken);
}
